File: explicit-java/src/main/kotlin/explicit/conversion/binding/ExplicitConversionBinding.kt
package explicit.conversion.binding

import explicit.api.IConversionBinding
import explicit.api.extension.removeSurroundingNonNumeric
import explicit.api.extension.toInt
import explicit.api.extension.toLong
import explicit.api.extension.toNumber
import java.math.RoundingMode
import java.time.LocalDateTime
import java.time.temporal.TemporalAdjusters
import java.util.*

class ExplicitConversionBinding(variables: Map<String, Any> = mapOf()) : IConversionBinding {

    private val variables = HashMap(variables)
    private val functions = HashMap<String, (List<Any>) -> Any>()

    init {
        functions["add"] = { it[0].toNumber().add(it[1].toNumber()) }
        functions["sub"] = { it[0].toNumber().subtract(it[1].toNumber()) }
        functions["mul"] = { it[0].toNumber().multiply(it[1].toNumber()) }
        functions["div"] = { it[0].toNumber().divide(it[1].toNumber(), RoundingMode.FLOOR) }
        functions["uppercase"] = { it[0].toString().toUpperCase() }
        functions["lowercase"] = { it[0].toString().toLowerCase() }
        functions["toNumber"] = { it[0].toNumber() }
        functions["first"] = { it.first { e -> e !is Optional<*> } }
        functions["ternary"] = { if (it[0].toString().toBoolean()) it[1] else it[2] }
        functions["isPresent"] = { variables.containsKey(it[0].toString()) }
        functions["date"] = {
            LocalDateTime.of(it[2].toInt(), it[1].toInt(), it[0].toInt(), 0, 0)
        }
        functions["addWeeks"] = { (it[0] as LocalDateTime).plusWeeks(it[1].toLong()) }
        functions["toDate"] = { args ->
            if (args.size == 2) {
                val delimiter = if (args[0].toString().contains("/")) "/" else "."
                val array = args[0].toString().split(delimiter)

                var year = ((array[2]).removeSurroundingNonNumeric()!!.toInt())
                year = if (year < 100) year + 2000 else year

                LocalDateTime.of(year, array[1].toInt(), array[0].removeSurroundingNonNumeric()!!.toInt(), 0, 0)
            } else {
                val delimiter = if (args[0].toString().contains("/")) "/" else "."
                val array = args[0].toString().split(delimiter)

                var year = args[2].toInt()
                year = if (year < 100) year + 2000 else year

                LocalDateTime.of(year, array[1].toInt(), array[0].removeSurroundingNonNumeric()!!.toInt(), 0, 0)
            }
        }
        functions["lastDayOfMonth"] = {
            val date = LocalDateTime.of(it[1].toInt(), it[0].toInt(), 1, 0, 0).with(TemporalAdjusters.lastDayOfMonth())
            date.dayOfMonth
        }
        functions["getYear"] = { it -> (it[0] as LocalDateTime).year }
        functions["curMonth"] = {LocalDateTime.now().month.value}
        functions["substringAfter"] = {it[0].toString().substringAfter(it[1].toString())}
    }

    override fun registerVariable(name: String, value: Any) {
        variables[name] = value
    }

    override fun registerFunction(name: String, function: (args: List<Any>) -> Any) {
        functions[name] = function
    }

    override fun getVariable(name: String) = variables[name]

    override fun getFunction(name: String): (List<Any>) -> Any {
        if (!functions.containsKey(name))
            throw RuntimeException("no function $name found")
        return functions[name]!!
    }
}

File: explicit-python/explicit_nlu/extractor/binding/ExplicitConversionBinding.py
import calendar
from datetime import date, timedelta
from datetime import datetime
from typing import Dict, Any, Callable, List

from explicit_nlu.api import IConversionBinding
from explicit_nlu.api.support import Strings


def to_date(args):
    if args[1] == "DMY":
        array = str(args[0]).split("." if "." in args[0] else "/")
        return datetime(int(array[2]) + 2000 if len(array[2]) == 2 else int(array[2]), int(array[1]), int(array[0]))
    if args[1] == "DM":
        array = str(args[0]).split(".")
        return datetime(int(args[2]), int(array[1]), int(array[0]))
    raise ValueError("cannot parse " + str(args) + " to date")


class ExplicitConversionBinding(IConversionBinding):

    def __init__(self, variables: Dict[str, Any]):
        self.variables = dict(variables)
        self.functions: Dict[str, Callable[[List[Any]], Any]] = {
            "add": lambda args: float(args[0]) + float(args[1]),
            "sub": lambda args: float(args[0]) - float(args[1]),
            "mul": lambda args: float(args[0]) * float(args[1]),
            "div": lambda args: float(args[0]) / float(args[1]),
            "uppercase": lambda args: str(args[0]).upper(),
            "lowercase": lambda args: str(args[0]).lower(),
            "toNumber": lambda args: Strings.to_number(args[0]),
            "first": lambda args: next(x for x in args if not x is None),
            "ternary": lambda args: args[1] if args[0] else args[2],
            "isPresent": lambda args: args[0] in self.variables,
            "date": lambda args: datetime(int(args[2]), int(args[1]), int(args[0])),
            "addWeeks": lambda args: args[0] + timedelta(days=int(args[1]) * 7),
            "lastDayOfMonth": lambda args: calendar.monthrange(int(args[1]), int(args[0]))[1],
            "getYear": lambda args: args[0].year,
            "curMonth": lambda args: date.today().month,
            "curDate": lambda args: date.today(),
            "substringAfter": lambda args: args[0][args[0].index(args[1]) + len(args[1]):],
            "toDate": to_date,
            "removeWhitespace": lambda args: args[0].replace(" ", ""),
            "replace": lambda args: args[0].replace(args[1], args[2]),
            "concat": lambda args: "".join(args),
        }

    def register_variable(self, name: str, value: Any):
        self.variables[name] = value

    def register_function(self, name: str, function: Callable[[List[Any]], Any]):
        self.functions[name] = function

    def get_variable(self, name: str):
        return self.variables[name] if name in self.variables else None

    def get_function(self, name: str) -> Callable[[List[Any]], Any]:
        if name in self.functions:
            return self.functions[name]
        raise ValueError("no function '" + name + "' found")
